src/Swarm/DocGen.hs

```diff
@@ -0,0 +1,247 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTSyntax #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+
+module Swarm.DocGen (
+  generateDocs,
+  GenerateDocs (..),
+) where
+
+import Control.Lens (view, (^.))
+import Control.Monad (zipWithM, zipWithM_, (<=<))
+import Control.Monad.Except (ExceptT, runExceptT)
+import Data.Bifunctor (Bifunctor (bimap), second)
+import Data.Containers.ListUtils (nubOrd)
+import Data.Foldable (toList)
+import Data.Map.Lazy (Map)
+import Data.Map.Lazy qualified as Map
+import Data.Maybe (fromMaybe)
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.Text (Text, unpack)
+import Data.Tuple (swap)
+import Swarm.Game.Entity (Count, Entity, EntityMap (entitiesByName), entityName, loadEntities)
+import Swarm.Game.Entity qualified as E
+import Swarm.Game.Recipe (Recipe, loadRecipes, recipeInputs, recipeOutputs, recipeRequirements)
+import Swarm.Game.Robot (installedDevices, robotInventory, setRobotID)
+import Swarm.Game.Scenario (Scenario, loadScenario, scenarioRobots)
+import Swarm.Game.WorldGen (testWorld2Entites)
+import Swarm.Util (isRightOr)
+import Text.Dot (Dot, NodeId)
+import Text.Dot qualified as Dot
+
+-- ============================================================================
+-- MAIN ENTRYPOINT TO CLI DOCUMENTATION GENERATOR
+-- ============================================================================
+--
+-- These are the exported functions used by the executable.
+--
+-- ----------------------------------------------------------------------------
+
+data GenerateDocs where
+  -- | Entity dependencies by recipes.
+  RecipeGraph :: GenerateDocs
+  deriving (Eq, Show)
+
+generateDocs :: GenerateDocs -> IO ()
```

Comment:
Perhaps we only need one command to generate the documentation files directly. I guess we could also render a markdown reference file, and it might be easier to just add `swarm generate-doc` in the ci, and fail if the output changed.
For the final dot render, I think it is not deterministic, so updating the image could be a manual step, though we could commit the dot source file.

Reply:
We could generate all the documentation and put it in appropriate files.
I can imagine a `--dir` option to `generate` that would create the files in the chosen directory.

I like the default `stdout` output for debugging and it also does not require the user to be in the swarm repo.

---

Eventually, I would like to create the render in CI and add it to GitHub Webpage.

The editor updates suggested in #344 seem more useful to track in this repository to me.

But we can set up the CI to push to GitHub Wiki or GitHub Webpage and track the generated stuff there.